examples/chef/common/chef-fan-control-manager.cpp

```diff
@@ -0,0 +1,178 @@
+    if (!speedSetting.IsNull())
+    {
+        ret = speedSetting.Value();
+    }
+
+    return aEncoder.Encode(ret);
+}
+
+Status ChefFanControlManager::HandleStep(StepDirectionEnum aDirection, bool aWrap, bool aLowestOff)
+{
+    ChipLogProgress(NotSpecified, "ChefFanControlManager::HandleStep aDirection %d, aWrap %d, aLowestOff %d",
+                    to_underlying(aDirection), aWrap, aLowestOff);
+
+    VerifyOrReturnError(aDirection != StepDirectionEnum::kUnknownEnumValue, Status::InvalidCommand);
+
+    uint8_t speedMax;
+    SpeedMax::Get(mEndpoint, &speedMax);
+
+    DataModel::Nullable<uint8_t> speedSetting;
+    SpeedSetting::Get(mEndpoint, speedSetting);
+
+    uint8_t newSpeedSetting = speedSetting.IsNull() ? 0 : speedSetting.Value();
+
+    if (aDirection == StepDirectionEnum::kIncrease)
+    {
+        if (speedSetting.IsNull())
+        {
+            newSpeedSetting = 1;
+        }
+        else if (speedSetting.Value() < speedMax)
+        {
+            newSpeedSetting = static_cast<uint8_t>(speedSetting.Value() + 1);
+        }
+        else if (speedSetting.Value() == speedMax)
+        {
+            if (aWrap)
+            {
+                newSpeedSetting = aLowestOff ? 0 : 1;
+            }
+        }
+    }
+    else if (aDirection == StepDirectionEnum::kDecrease)
+    {
+        if (speedSetting.IsNull())
+        {
+            newSpeedSetting = aLowestOff ? 0 : 1;
+        }
+        else if ((speedSetting.Value() > 1) && (speedSetting.Value() <= speedMax))
+        {
+            newSpeedSetting = static_cast<uint8_t>(speedSetting.Value() - 1);
+        }
+        else if (speedSetting.Value() == 1)
+        {
+            if (aLowestOff)
+            {
+                newSpeedSetting = static_cast<uint8_t>(speedSetting.Value() - 1);
+            }
+            else if (aWrap)
+            {
+                newSpeedSetting = speedMax;
+            }
+        }
+        else if (speedSetting.Value() == 0)
+        {
+            if (aWrap)
+            {
+                newSpeedSetting = speedMax;
+            }
+            else if (!aLowestOff)
+            {
+                newSpeedSetting = 1;
+            }
+        }
+    }
+
+    return ToInteractionModelStatus(SpeedSetting::Set(mEndpoint, newSpeedSetting));
+}
+
+CHIP_ERROR ChefFanControlManager::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == FanControl::Id);
+    VerifyOrDie(aPath.mEndpointId == mEndpoint);
+
+    switch (aPath.mAttributeId)
+    {
+    case SpeedCurrent::Id:
+        return ReadSpeedCurrent(aEncoder);
+    case PercentCurrent::Id:
+        return ReadPercentCurrent(aEncoder);
+    default:
+        break;
+    }
+    return CHIP_NO_ERROR;
+}
+
+} // anonymous namespace
+
+void emberAfFanControlClusterInitCallback(EndpointId endpoint)
+{
+    VerifyOrDie(mFanControlManager == nullptr);
+    mFanControlManager = new ChefFanControlManager(endpoint);
```

Comment:
do we care that this never gets detete[] ? should we std::unique_ptr on it so that it attempts to be deleted when the program exits (so that any leak checkers we run would not complain)?

Reply:
I don't think we have the callback when Cluster is deinitialized

Reply:
if you wrap it in a std::unique_ptr at least if we ever run a leak checker the memory will be freed at program exit. That is better than a raw pointer.

Reply:
OK. Pushed.